internal/pkg/client/patch/patch.go

```diff
@@ -0,0 +1,210 @@
+// The implementation here has been mostly extracted from the apply command: k8s.io/kubernetes/pkg/kubectl/cmd/apply.go
+func GetClientSideApplyPatch(current, desired runtime.Object) (Patch, error) {
+	var patch Patch
+	// Serialize the current configuration of the object.
+	currentb, err := runtime.Encode(unstructured.UnstructuredJSONScheme, current)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize current %v", current)
+	}
+	// Retrieve the last applied configuration of the object from the annotation.
+	last, err := GetLastApplied(current)
+	if err != nil {
+		return patch, fmt.Errorf("could not retrieve original configuration from %v", desired)
+	}
+
+	// Serialize the modified configuration of the object, populating the last applied annotation as well.
+	modified, err := GetModifiedConfiguration(desired, true)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize intended configuration from %v", desired)
+	}
+	gvk := desired.GetObjectKind().GroupVersionKind()
+
+	versionedObject, err := Scheme.New(gvk)
+	_, unversioned := Scheme.IsUnversioned(desired)
+	switch {
+	case runtime.IsNotRegisteredError(err) || unversioned:
+		preconditions := []mergepatch.PreconditionFunc{
+			mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"),
+			mergepatch.RequireMetadataKeyUnchanged("name"),
+		}
+		patch.Data, err = jsonmergepatch.CreateThreeWayJSONMergePatch(last, modified, currentb, preconditions...)
+		patch.Type = types.MergePatchType
+		if err != nil {
+			name, err := metadataAccessor.Name(desired)
+			if mergepatch.IsPreconditionFailed(err) {
+				return patch, fmt.Errorf("at least one of apiVersion, kind and name was changed for %s/%s", gvk, name)
+			}
+			return patch, err
+		}
+	case err != nil:
+		return patch, err
+	case err == nil:
+		// Compute a three way strategic merge patch to send to server.
+		patch.Type = types.StrategicMergePatchType
+		lookupPatchMeta, err := strategicpatch.NewPatchMetaFromStruct(versionedObject)
+		if err != nil {
+			return patch, err
+		}
+		patch.Data, err = strategicpatch.CreateThreeWayMergePatch(last, modified, currentb, lookupPatchMeta, true)
+		if err != nil {
+			return patch, err
+		}
+	}
+	return patch, nil
+}
+
+// GetMergePatch - generate merge patch from original and modified objects
+func GetMergePatch(original, modified runtime.Object) (Patch, error) {
+	originalJSON, err := json.Marshal(original)
+	if err != nil {
+		return Patch{}, err
+	}
+
+	modifiedJSON, err := json.Marshal(modified)
+	if err != nil {
+		return Patch{}, err
+	}
+
+	patch, err := jsonpatch.CreateMergePatch(originalJSON, modifiedJSON)
+	if err != nil {
+		return Patch{}, err
+	}
+	return Patch{types.MergePatchType, patch}, nil
+}
+
+// GetLastApplied retrieves the original configuration of the object
+// from the annotation, or nil if no annotation was found.
+func GetLastApplied(obj runtime.Object) ([]byte, error) {
```

Comment:
lets not export these functions if they aren't used outside the package.

Reply:
This is being used in test code. Alternative is to implement the same in test code.